Facet API upgrade for backdrop

earlyburg opened the pull request:
remove ctools functions
convert drupal_somefunctions() to backdrop_somefunctions()
convert and repair block-caching functionality
remove old caching functionality
create new config db table and update hook
create new dynamic variable storage function
remove redundant functions
remove comment
add README.md
add LICENCE.txt

earlyburg committed:
convert ctools functionality re. new functions
remove ctools functions
convert drupal_somefunctions() to backdrop_somefunctions()
convert and repair block-caching functionality
remove old caching functionality
create new config db table and update hook
create new dynamic variable storage function
remove redundant functions
remove comment
add README.md
add LICENCE.txt

earlyburg commented:
I have done my best to complete as per your suggestions. A lot of them I have agreed with. If I have omitted anything I will leave it up to you to improve my work as you see fit. Again, I appreciate your assistance with this. Thank you.

docwilmot commented on facetapi.install:
```diff
@@ -73,30 +73,36 @@ function facetapi_schema() {
     'primary key' => array('name'),
   );
-
+  $schema['facetapi_config'] = array(
```
I'm still not understanding why youre creating this table instead of storing these variables in CMI.

earlyburg commented:
@docwilmot 
From my limited understanding of backdrops CMI:

function config_set($config_file, $option, $value) {
  $config = config($config_file);
  $config->set($option, $value);
  $config->save();
}

I would like **$option** to be created dynamically and saved, and I am under the impression that this config variable name has to be existing already in the modulename.settings.json file.

So if I wanted to store config json like:
dynamicname4: '20',
dinamicframeX11: 'ABC',
bionicblame: 'total'
 
then CMI would not work? Again I have limited understanding of the capabilities but I did read the docs :)

I'm mindful that in D7 variable_get() will set a new variable of any name value pair and return it. 

function variable_get($name, $default = NULL) {
  global $conf;
  return isset($conf[$name]) ? $conf[$name] : $default;
}

If you could kick me an example of how this would be done...
Thanks for your patience with this issue, I appreciate it.

docwilmot commented:
You can do exactly the same as you would with `variable_set()`: for example:
   ` config_set('facetapi.settings', 'facetapi:block_cache:' . $searcher, MYVALUE)`

The difference with `variable_get()` is that it allows a default value if `facetapi:block_cache:' . $searcher` was not previously set. How we fix this in Backdrop is to ensure that we have written default values in the default config file:
```
     "facetapi:block_cache:searcher_name1": "value",
     "facetapi:block_cache:searcher_name2": "value2",
```
Or if we cannot be sure about searcher name until install, we can create these values in an install hook:
```
  function facetapi_install() {
    SEARCHERNAMES = method_to_get_all_searcher_names();
    foreach (SEARCHERNAMES as SEARCHERNAME) {
      config_set('facetapi.settings', 'facetapi:block_cache:' . SEARCHERNAME, MYVALUE);
    }
  }
```

Note of course this is example pseudocode.

earlyburg commented:
@docwilmot 
What about setting SEARCHERNAME every time a search index specific setting is set?
There can be multiple search indexes accessed by facet api and I am attempting to set their configuration varaibles.
I can't know these until after module install.
The issue is that we can not know the SEARCHERNAME until a facet relationship is configured. After the install.

I want something analogous to:
$dynamicname = array('myUniqueName1', 'myUniqueName2', 'myUniqueName3', );
foreach($dynamicname as $value) {
   config_set('facetapi.settings', 'facetapi: . $value . NAME:' , 'SOMEVALUE')
}

earlyburg commented:
@laryn apologies wrong pull request

docwilmot commented:
You'd do the same as in the example above. From the existing code:
In a submit function you changed this:
```
    $name = 'facetapi:block_cache:' . $adapter->getSearcher();
    $original = variable_get($name, BACKDROP_NO_CACHE);
    variable_set($name, $form_state['values']['block_cache']);
```

To this:
```
    $name = 'facetapi:block_cache:' . $adapter->getSearcher();
    $original = facetapi_config($name);
    facetapi_config($name, $form_state['values']['block_cache']);
```

The Backdrop equivalent would simply be:
```
    $name = 'facetapi:block_cache:' . $adapter->getSearcher();
    $original = config_get('facetapi.settings', $name);
    config_set('facetapi.settings', $name, $form_state['values']['block_cache']);
```

If there were multiple adapters that existed in a freesh install (I dont know, I am not familiar with Facetapi, this is just an example), and you wanted the forms to have default values on install, you would do something like this in the install file:
```
  function facetapi_install() {
    $adapters = method_to_get_all_adapters();
    foreach ($adapters as $adapter) {
      config_set('facetapi.settings', 'facetapi:block_cache:' . $adapter->getSearcher()->name, MYVALUE);
    }
  }
```

earlyburg commented:
> If there were multiple adapters that existed in a freesh install (I dont know, I am not familiar with Facetapi, this is just an example), 

@docwilmot
The condition we are trying to address is that after the install, multiple adapter names are created and must be saved.
The code creates the **name** of the name => value pair when any search index is referenced. 
Again, my question is simply how to set a configuration variable **name** and value after install?

The original line is:

$someValue = variable_get('facetapi:block_cache:' . $searcher, DRUPAL_NO_CACHE),

In this instance there is no pre-existing variable name that a value is associated with.
The name and value are applied directly as a new configuration setting, and then returned.

If there is no way to know the variable name, I can't assign it in an install hook.
If I can't reference each newly created search index identifer then every search index created after the install of facetapi could not use facets for search result filtering.

Adapters are being added and removed all the time when a new search index is created or removed.
With the json system, it's definitely cleaner but it's not possible to actually _write additional json_ to the file, only to assign variables based on the preconfigured json file. Does that make sense?
Thanks for your patience. I just don't see in any of your examples (which are awesome) the functionality that we need.

docwilmot commented:
> it's not possible to actually write additional json to the file, only to assign variables based on the preconfigured json file

I don't understand what this means. Maybe we can private chat on Zulip to resolve this.

earlyburg commented:
Do what thou wilt...
I think I might try using an array as my value for the config setting.

config_set('facetapi.settings', 'index_caches', []);
then go
config_set('facetapi.settings', 'index_caches', [ 'searchIndex1': 'facet_0_Name', 'searchIndex2': 'facet_2_Name',]);

could I then go:
$value = json_decode(config_get('facetapi.settings', 'index_caches')); 
foreach ($value as $val) {
  if ($val == $searchVar) {
     // return the value
  }
}
?
I have to make a test case though, and I'm currently back on another project.
For the record - the current pull request works flawlessly, although not "ideally".
Half-full from my perspective.

earlyburg committed:
Remove config var block_cache:search and references
Remove hook_block_info() cache setting entries
Remove unwanted function references
Remove ctools function and references
Remove block cache setting form element
Remove unneeded schema and update hook

earlyburg commented:
@docwilmot I have completed the changes that we discussed. No more block caching.
Thanks for your paitence with this I appreciate it.

docwilmot commented:
I'm merging this so we can update the maintainer listing, but I haven't really reviewed the code.
